src/Microsoft.OData.Core/JsonLight/ODataJsonLightContextUriParser.cs

```diff
@@ -61,27 +61,30 @@ private ODataJsonLightContextUriParser(IEdmModel model, Uri contextUriFromPayloa
         /// <param name="model">The model to use when resolving the target of the URI.</param>
         /// <param name="contextUriFromPayload">The string value of the odata.metadata annotation read from the payload.</param>
         /// <param name="payloadKind">The payload kind we expect the context URI to conform to.</param>
-        /// <param name="clientCustomTypeResolver">The function of client cuetom type resolver.</param>
+        /// <param name="clientCustomTypeResolver">The function of client custom type resolver.</param>
         /// <param name="needParseFragment">Whether the fragment after $metadata should be parsed, if set to false, only MetadataDocumentUri is parsed.</param>
         /// <param name="throwIfMetadataConflict">Whether to throw if a type specified in the ContextUri is not found in metadata.</param>
+        /// <param name="baseUri">Optional value (with default value of null) of base Uri used for resolving the context url.
+        /// It should be a non-null value when resolving a top-level relative context Uri.
+        /// </param>
         /// <returns>The result from parsing the context URI.</returns>
         internal static ODataJsonLightContextUriParseResult Parse(
             IEdmModel model,
             string contextUriFromPayload,
             ODataPayloadKind payloadKind,
             Func<IEdmType, string, IEdmType> clientCustomTypeResolver,
             bool needParseFragment,
-            bool throwIfMetadataConflict = true)
+            bool throwIfMetadataConflict = true,
+            Uri baseUri = null)
         {
             if (contextUriFromPayload == null)
             {
                 throw new ODataException(ODataErrorStrings.ODataJsonLightContextUriParser_NullMetadataDocumentUri);
             }
 
             // Create an absolute URI from the payload string
-            // TODO: Support relative context uri and resolving other relative uris
             Uri contextUri;
-            if (!Uri.TryCreate(contextUriFromPayload, UriKind.Absolute, out contextUri))
+            if (!Uri.TryCreate(baseUri, contextUriFromPayload, out contextUri))

```

Comment:
What will happen if contextUriFromPayload is absolute? I'm reading the [API here](https://msdn.microsoft.com/en-us/library/ms131573(v=vs.110).aspx) and it says that the second Uri is relative.

Reply:
I believe it might be some doc issue. The actual implementation of Uri.TryCreate(Uri, string, out Uri) first tries to convert the string to absolution uri first if possible; otherwise it tries to use the baseUri to resolve the string representing relative url to get the result. See here: [https://github.com/Microsoft/referencesource/blob/master/System/net/System/UriExt.cs#L315](https://github.com/Microsoft/referencesource/blob/master/System/net/System/UriExt.cs#L315)


---
In reply to: [202220739](https://github.com/OData/odata.net/pull/1211#discussion_r202220739) [](ancestors = 202220739)